server/graphql/programs/src/mutations/encounter/insert.rs

```diff
@@ -45,6 +45,14 @@ pub fn insert_encounter(
     let service_provider = ctx.service_provider();
     let service_context = service_provider.basic_context()?;
 
+    match user.context.into_iter().find(|c| c == &input.r#type) {
```

Comment:
this sort of mapping is usually done with [ok_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or), we use it in service validations and in some places in graphql

Reply:
Hmm, so the context in this case would be any context (both mutate and query) from all of the user permissions, there are a number of ways to solve this:
* Pass context through to validation through a string enum variant: `Resource::MutateEncounter(String)`, and make `HasPermissionWithContext { permisison, context }`, but existing DSL would need to be refactored quite a bit to pass `context` from MutateEncounter `Resource` variant to HasPermissionWithContext variant of `PermissionDSL` : /
* Another way is for `validate_resource_permissions` to return array of context that match permission type, but because validate_resource_permissions is nested with possible `AND` or `OR` being the parent, it seems tricky and hacky and would only work for one uses case where there is no OR and AND in permission DSL variants.

I would make and issue with permission improvements shortly, it would question the need for permission DSL and how much complexity are we willing to accept by continuing to squeeze functionality into it. I would prefer for this to be a very simple validation done in `validate` method in service (with some helpers)

Like this: https://github.com/openmsupply/open-msupply/compare/29-Document-permissions...29-Validation-in-service

Reply:
The example with validation in the service layer looks nice to me.
I think it could still be combined with DSL If needed for a particular use case. No reason (that I'm aware of) that the DSL is limited to Graphql layer.

Reply:
I read this too late and already toyed around with option 2. The `AND` / `OR` cases where actually not too bad to handle and I made `validate_resource_permissions` return list of permissions needed for this endpoint. However, its still a bit tricky to use and after discussion with @andreievg we thought it might be easier to have a helper method that is more straight forward. 

Regarding, service validation. I like the separation of concern between user permissions and services, i.e. the permission system is independent from the service logic. This makes it possible to use the service in other places without running into permission issues, e.g. a cleanup, sync or migration task can use the service. Moreover, we could make changes to the user permissions without affecting the services. Its also not too nice to have to read the user permissions in two different places, i.e. in the DSL and in a service.

However, I think we can find a good compromise here. I would keep reading and handling the user permission in the graphql layer next to the DSL validation and extract a list of relevant contexts using a helper method. This list is then passed on to the service and the service can do it's validation. For example,
```rust
// in graphql:
let validated_user = validate(...)?;
let allowed_doc_types: Vec<String> = helper_get_permissions_contexts(Permission::DocumentEncounterQuery, validated_user.user_permissions);

service.query(allowed_doc_types, ...)
```

In other words we extract the user's capabilities from the user permissions and pass them on to the service and the service can do its validation based on these capabilities.